### PR TITLE
Clean-up and refactoring for code transitioning over to graphql-dotnet/conventions

### PR DESCRIPTION
**Pull request by tlil87**
#### Changes
 * `GetArgument()` wasn't using the `FieldDefinition`'s default value for an argument correctly. Added method signature for using the `FieldDefinition` default, and one for overriding the value
 * `ConnectionType<T>`, `EdgeType<T>` and `PageInfoType` implementations have been separated from their backing data objects. Improved and more consistent Relay-specific helper functions for these will be transitioned over to graphql-dotnet/conventions
 * Moved Relay-specific type structures to the `GraphQL.Types.Relay` namespace
 * Removed `new()` constraint on connections and edges
 * Exposing `ResolveFieldContext` in `FieldBuilder` and `ConnectionBuilder` (and consequently also the `CancellationToken`, etc.)
 * Added new scalar type `DecimalGraphType`
 * Added support for deprecation of `EnumerationGraphType` members
 * Added a range of unit tests for the above
 * Fixed `IdGraphType.Coerce()`
 * `IntGraphType.Coerce()` can now also deal with `long` integers

### CODE REVIEW

**Comment by joemcbride** [comment]
I haven't had a chance to dig into this yet, though I'll see if I can get to it over the weekend.

I spent entirely too much time investigating a switch to xunit since I thought that would help with some of the build server issues.  It didn't.  I did finally figure out how to properly get it to fail a build with fixie though.

In short - I (and everyone else) should be able to trust AppVeyor for branches that are off of the latest master.

File: src/GraphQL/Execution/DocumentExecuter.cs [hunk 1]
```diff
@@ -373,6 +373,18 @@ public Variables GetVariableValues(ISchema schema, Variables variables, Inputs i
...
+            if (type == null && !variable.Type.AllowsNull)
+            {
+                var nonNullType = new VariableType
+                {
+                    Name = variable.Type.Name,
+                    IsList = variable.Type.IsList,
+                    AllowsNull = true,
+                };
+                type = schema.FindType(nonNullType.FullName);
+            }
```
Comment (joemcbride) [inline src/GraphQL/Execution/DocumentExecuter.cs]: Do you have an example of this failing and where it is needed?  This seems like an odd and incorrect change.  "If the type can't be found and its non-null, create a nullable variable type".  Is the issue the name is coming through as `MyType!` and the lookup only knows about `MyType`?

File: src/GraphQL/Types/GraphTypesLookup.cs [hunk 2]
```diff
@@ -118,9 +118,9 @@ public void AddType(GraphType type, TypeCollectionContext context)
                 }
             });
 
-            if (type is ObjectGraphType)
+            var obj = type as ObjectGraphType;
+            if (obj != null)
```
Comment (joemcbride) [inline src/GraphQL/Types/GraphTypesLookup.cs]: The pattern here was the direct casts so variables didn't leak into the outer scope  - as is such the case now with this change.

Comment (tlil87) [inline src/GraphQL/Types/GraphTypesLookup.cs]: Gotcha, will change that back

File: src/GraphQL/Execution/DocumentExecuter.cs (hunk 1 above)
Comment (tlil87) [inline src/GraphQL/Execution/DocumentExecuter.cs]: Sorry, forgot to take this out. You're absolutely right. Removing.

**Comment by tlil87** [comment]
@MartinLorensen does this branch resolve the issues that you were seeing before?

**Comment by MartinLorensen** [comment]
@tlil87: Du you refere to the issues with unquoting in IdGraphType.Coerce() ? I have just verified that I can remove my `id = id.Trim('\"')` from my node-field handling - so if that was the issue I would say it has been solved.

**Comment by tlil87** [comment]
@MartinLorensen yep, was referring to that and the `new()` constraint that we got fixed earlier :+1: thanks for confirming

**Comment by MartinLorensen** [comment]
@tlil87 Ahh - okay. And yes, as indicated in that issue, it was solved too.

**Pull request merged**